### ISSUE TITLE
Digest State Serialization

### ISSUE THREAD

**Issue by slontis** [openssl/openssl#14222]
Functionality to load and save the internal state of a digest would be useful for systems using multi processors with fail-over capabilities. This would allow the user to save a blob of data any time after the init has completed.

A possible implementation would need to do the following:
 - High level API's to serialize and deserialize an EVP_MD_CTX.
 - Each provider digest implementation would need to have methods to serialize/deserialize their internal data.
 - Have some blob format. 
 
Some issues related to the blob format (if the data was not ephemeral):
- Handling backwards compatibility would complicate the serialization (since any small change to a digest would cause compatibility problems). Any additional new fields in a digest struct could potentially behave differently and have no way of working with older code.
- A different provider could possibly serialize a digest differently.
- A different implementation (e.g uses different assembler) could serialize differently.

**Comment by paulidale** [comment]
Why isn't the existing `EVP_MD_CTX_dup()` adequate?  It saves the current state and can subsequently be used.
What it doesn't permit is migration between machines or being saved to file, etc.

**Comment by slontis** [comment]
The use case is using Shared Memory so dup would not work for this..

**Comment by paulidale** [comment]
What's wrong with duping into shared memory by hooking the memory allocation calls?

This is, by necessity, going to be version and provider specific.  It also feels like a hack.

**Comment by slontis** [comment]
The alloc hooks would put 'everything' into shared memory which would not be ideal..

**Comment by slontis** [comment]
I think it would need to assume that the data is ephemeral..

**Comment by paulidale** [comment]
Why not:

``` C
thread_local alloc_in_shared_memory = 0
...
alloc_in_shared_memory = 1
EVP_MD_CTX_dup(...)
alloc_in_shared_memory = 0
```

**Comment by slontis** [comment]
The free would know which one it was based on the address range?

**Comment by paulidale** [comment]
Exactly.   If you want to migrate something, you should already know you want to and can distinguish. If not, wrap it in a structure and have a flag.

**Comment by t8m** [comment]
Digest (and possibly other) state serialization would be useful in many ways however IMO the only remotely sane way is to make it completely provider-specific internal thing. And the provider implementation would have to add some uuid to the serialized data so it could be checked for compatibility on deserialization. The other way would be to explicitly define the serialization data format for each algorithm but that would be a huuge task.

**Comment by paulidale** [comment]
It's more than just provider side: I think libcrypto should add enough information to recreate the context -- provider name, algorithm name, selection properties and maybe more.

I agree, it would be a big job.  How much value does it add beyond the existing dup functions?

**Comment by t8m** [comment]
IMO it would be valuable enough to do it at some point. But the question is when is that point - perhaps after we've done QUIC, FIPS 140-3, PKCS#11 provider, removal of legacy... then it might be something to tackle. :laughing:

**Comment by richsalz** [comment]
`alloc_in_shared_memory = 1` is that a global variable?  how does it work in multi-threaded programs?

**Comment by paulidale** [comment]
> `alloc_in_shared_memory = 1` is that a global variable? how does it work in multi-threaded programs?

It's declared with a `thread_local` prefix, so I guess not.

**Comment by richsalz** [comment]
> It's declared with a thread_local prefix, so I guess not.

So windows-only I'm guessing.  `alloc_in_shared_memory` isn't in OpenSSL sources and I know Windows has that nice `shared` keyword. :)

**Comment by paulidale** [comment]
It was just an example of _how_ it could be done without this feature.  I didn't write the custom memory handlers...

**Comment by ifranzki** [comment]
That's definitely something that I also need!

With OpenSSL 1.1.1 I do this currently the following way:

To save the digest state of an active digest context:
- Use `EVP_MD_meth_get_app_datasize()` to get the length of the state data.
- Use `EVP_MD_CTX_md_data()` to get the address of the context internal state data.
- Copy the state data to some place to keep the state.

To restore the state later on:
- Create a new digest context with `EVP_DigestInit_ex()`.
- Use `EVP_MD_meth_get_app_datasize()` to get the length of the state data.
- Use `EVP_MD_CTX_md_data()` to get the address of the context internal state data.
- Copy the to be restored state to the address obtained from `EVP_MD_CTX_md_data()`.
- Use the digest context, i.e. do `EVP_DigestUpdate()` etc.

This works well on OpenSSL 1.1.1, but with OpenSSL 3.0 `EVP_MD_meth_get_app_datasize()` is deprecated.
`EVP_MD_CTX_md_data()` is now `EVP_MD_CTX_get0_md_data()`, but is still available with OpenSSL 3.0 .
However, without knowing the size of the state data it is pretty useless. So this leaves me in a situation where an application that works with OpenSSL 1.1.1 won't work anymore, once the deprecated functions will be gone. 

I would guess that the OpenSSL 3.0 way of doing this would be to use `EVP_MD_CTX_get_params()` and `EVP_MD_CTX_set_params()` with new parameters like OSSL_DIGEST_PARAM_STATE_SIZE and OSSL_DIGEST_PARAM_STATE_DATA. One could obtain the current state data of a digest context and its size using `EVP_MD_CTX_get_params()`. To restore the state on an exiting digest context `EVP_MD_CTX_set_params()` could be used, or even quicker, `EVP_DigestInit_ex2()` with specifying params containing the to be restored state to create a new digest context with the desired state.

Since the set_params/get_params calls are anyway passed to the provider, it should be relatively easy to implement that within a provider.  No API change needed, except the new OSSL_DIGEST_PARAM's. 

The provider could or should probably embed the digest type and used provider within the serialized state data, so that it can check if a user provided state data works for the selected digest and provider.

**Comment by paulidale** [comment]
You're only starting to touch on the difficulties involved in supporting this.  The structures in 3.0 are quite different and now contain pointers to provider side data.  Both need to be serialised.  That's in addition to remembering the provider etc.  Moreover, the provider data is an opaque blob too and for some algorithms there are pointers to other dynamically allocated things.

For the moment, the dup calls are the best we can offer.


Even if you were willing to use the deprecated 1.1.1 calls, it's not going to work.

**Comment by levitte** [comment]
`EVP_MD_CTX_md_data()` isn't, and shouldn't be relevant for provider interfaces.  It's a legacy construct.
The corresponding thing for provider implementations is the provider-side digest context, and should be expected to be entirely provider specific and provider private.  I cannot fathom why it should be serialized, as that implies it would be transferred to *other* providers.

**Comment by levitte** [comment]
Do note that the pointer returned by `EVP_MD_CTX_md_data()` was also a blob private to the backend that populates it.

**Comment by ifranzki** [comment]
> I cannot fathom why it should be serialized, as that implies it would be transferred to other providers.

Its not about transferring it to other provides, it is to to save the state and continue at a alter time (on the same provider).

Specifically, I need it for supporting a PKCS#11 implementation. PKCS#11 defines functions C_GetOperationState() and C_SetOperationState() that can save the state of an operation (i.e a digest operation in this case), and continue it at a later time.

I would be totally OK if it would be my own job to ensure that the saved state is being used on the same provider and digest type (or if it simply fails if it is not matching).

**Comment by ifranzki** [comment]
> Even if you were willing to use the deprecated 1.1.1 calls, it's not going to work.

That was my fear.... One more reason to have an official interface for doing such things in OpenSSL 3.0.

**Comment by ifranzki** [comment]
> ... is the provider-side digest context, and should be expected to be entirely provider specific and provider private. 

Right, that's why I think exposing it via get_params / set_params is the right way. Those get/set params are handled inside the provider, so a provider can serialize it in whatever way it likes to. The provider can handle pointers inside the state data as needed. Its all within the provider.

**Comment by levitte** [comment]
> Its not about transferring it to other provides, it is to to save the state and continue at a alter time (on the same provider).

In that case, `EVP_MD_CTX_dup()` should do the job just fine, no?  It holds a pointer to the provider-side context, so it should do exactly what you want.

> Specifically, I need it for supporting a PKCS#11 implementation. PKCS#11 defines functions C_GetOperationState() and C_SetOperationState() that can save the state of an operation (i.e a digest operation in this case), and continue it at a later time.

So am I getting it right, that you're trying to make a PKCS#11 provider?

> > ... is the provider-side digest context, and should be expected to be entirely provider specific and provider private.
> 
> Right, that's why I think exposing it via get_params / set_params is the right way.

No.  That would imply that you're transferring the state *out of the provider* to be used by whatever else, completely unknown to that provider's code.  That is contrary to the idea of "provider private".

**Comment by ifranzki** [comment]
> So am I getting it right, that you're trying to make a PKCS#11 provider?

No, I have a PKCS#11 implementation (openCryptoki) that uses OpenSSL crypto service to implement the PKCS'11 functions (software HSM so to speak).

**Comment by paulidale** [comment]
> In that case, EVP_MD_CTX_dup() should do the job just fine, no?

Dup() isn't sufficient.  It requires shared memory which isn't always available.


> No, I have a PKCS#11 implementation (openCryptoki) that uses OpenSSL crypto service to implement the PKCS'11 functions (software HSM so to speak).

dup might be sufficient for this.  dup the context, return that C_GetOperationState() and dup it again in C_SetOperationState().  It won't survive power cycles but, depending on the situation, might be enough.

**Comment by ifranzki** [comment]
Where is `EVP_MD_CTX_dup()` defined? I don't find it anywhere in the code....

**Comment by paulidale** [comment]
We've wrapped it up in `EVP_MD_CTX_copy_ex()`.  Ugh.

You'll be able to use `EVP_MD_CTX_copy_ex()` in the way I described, you'll just have to create your own context using `EVP_MD_CTX_new()`.

**Comment by ifranzki** [comment]
OK, `EVP_MD_CTX_copy_ex()` is there. But how does this help me to get the state in a way that I can pass it back via `C_GetOperationState()` ? What I need is a blob that contains the state. 

With EVP_MD_CTX_copy_ex() I can copy it to another EVP_MD_CTX, but how do I get the state data from that other context then? 

You mention alloc hooks, how does this work? And would that be really save, especially when the state data contains pointers?

**Comment by paulidale** [comment]
Doesn't this do something like what's required?

```c
const void *C_GetOperationState(const EVP_MD_CTX *mdctx) {
    EVP_MD_CTX *ret = EVP_MD_CTX_new();

    EVP_MD_CTX_copy_ex(ret, mdctx);
    return ret;
}

EVP_MD_CTX *C_SetOperationState(const void *state) {
    EVP_MD_CTX *ret = EVP_MD_CTX_new();

    EVP_MD_CTX_copy_ex(ret, (const EVP_MD_CTX *)state);
    return ret;
}
```

**Comment by ifranzki** [comment]
No, This is how C_GetOperationState is defined by the PKCS#11 standard:

CK_DECLARE_FUNCTION(CK_RV, C_GetOperationState)(
  CK_SESSION_HANDLE hSession,
  CK_BYTE_PTR pOperationState,
  CK_ULONG_PTR pulOperationStateLen
);

CK_DECLARE_FUNCTION(CK_RV, C_SetOperationState)(
  CK_SESSION_HANDLE hSession,
  CK_BYTE_PTR pOperationState,
  CK_ULONG ulOperationStateLen,
  CK_OBJECT_HANDLE hEncryptionKey,
  CK_OBJECT_HANDLE hAuthenticationKey
);

So the operation state is actually a blob of an arbitrary length, and has nothing directly to do with an OpenSSL MD context.

**Comment by levitte** [comment]
@ifranzki, are you trying to implement a PKCS#11 based provider, or are you trying to make a PKCS#11 implementation based on OpenSSL?

**Comment by levitte** [comment]
It appears to me that `pOperationState` is the pointer that would point at, well, something.  I don't understand how you're trying to make that interact with `EVP_MD_CTX` or the provider-side counterpart, or why a duplication with EVP_MD_CTX_copy_ex() isn't good enough.

**Comment by ifranzki** [comment]
> are you trying to implement a PKCS#11 based provider, or are you trying to make a PKCS#11 implementation based on OpenSSL?

Opencryptoki has a soft token, that is a PKCS#11 implementation based on OpenSSL.

> It appears to me that pOperationState is the pointer that would point at, well, something.

exactly. pOperationState is a pointer to a user supplied buffer, where C_GetOperationState is supposed to place the blob of the state into. pulOperationStateLen is an address of a ulong that on entry contains the length of the supplied buffer, and on return it is updated with the actual length of the state blob. 

> I don't understand how you're trying to make that interact with EVP_MD_CTX or the provider-side counterpart, or why a duplication with EVP_MD_CTX_copy_ex() isn't good enough.

That's exactly the problem I have. I need a way to get a **blob** of the current digest state so that I can copy it into the user supplied buffer in C_GetOperationState, and restore it later on in C_SetOperationState.
EVP_MD_CTX_copy_ex() dosn't give me a blob, it just gives me another OpenSSL digest context.

**Comment by t8m** [comment]
It is impossible to implement C_GetOperationState/C_SetOperationState properly without serialization unless you limit validity of the operation state to the running process - then you can just make the operation state to be the pointers to the EVP_MD_CTX.

Of course serialization would be a nice feature, but it is seriously hard to implement right and definitely won't happen in 3.0.

As for your code workarounding the need for proper serialization in 1.1.1 - this somewhat works but I would not hold my breath on what happens if for example you serialize a hash state and in a separate process after an upgrade of openssl to a new version (from the same branch even) you try to reuse the previously serialized state. It of course cannot work in 3.0 at all.

**Comment by levitte** [comment]
I just read up on `C_GetOperationState` / `C_SetOperationState`...  I hadn't realised that they can be used to pass state between different sessions.  Yeah, we have zero support for that in OpenSSL 3.0 (I don't think that the previous MD data was designed with this in mind either, so using that was pure hackery on someone's part).

I'm afraid that the current choice is for `C_GetOperationState` based on OpenSSL 3.0 to return `CKR_STATE_UNSAVEABLE`...

I suggest that we take this as food for thought as a future feature.

**Comment by ifranzki** [comment]
> I would not hold my breath on what happens if for example you serialize a hash state and in a separate process after an upgrade of openssl to a new version

Agreed, when the digest context structs change or compiler aligning changes then it will definitely fail. 

It is not stated that way in the PKCS#11 standard , but I think that C_GetOperationState/C_SetOperationState is mostly used in that way that you initialize the digest operation, maybe perform a first DigestUpdate on it, and then save the sate. You then continue with digesting additional data to get a final digest. Then you restore the state, and add other data to be digested. So you essentially re-use the digested data up to the point where you saved the state. 

So for this use case, I could probably just pass back the EVP_MD_CTX as is. But, the way the PKCS#11 API works, I can't pass back the pointer to EVP_MD_CTX, but I need to copy it into the user supplied buffer. However, I don't know the size of EVP_MD_CTX because its opaque....

Anyway, I would not be able to detect other use that restoring a state form within the same process, which then will most likely fail/segfault.... So I guess its not a good choice of doing it that way at all.
As you said, it can't be properly implemented without state serialization done by the providers.

>  It of course cannot work in 3.0 at all.

That was my fear ....

So I guess I have to handle it that way that I return CKR_STATE_UNSAVEABLE for those operations that involve digests, at least when compiled against OpenSSL 3.0.

Anyway: I guess you understood that digest state serialization is a very important feature for us. Maybe this helps to move it up a little on the list of features to be implemented after OpenSSL 3.0.....

**Comment by t-j-h** [comment]
> I just read up on `C_GetOperationState` / `C_SetOperationState`

And you have to keep in mind that `CKR_FUNCTION_NOT_SUPPORTED` is commonly returned when you call this function.
As is `CKR_STATE_UNSAVEABLE` for when there are limited circumstances where they are supported.
There are also restrictions on key state being exported, and the concept of requiring live handles to the keys to be provided when setting the operation state. 

These are explicitly not pure serialisation functions.

Implementations are not required to support this capability.

If you look at the [Baseline Provider Clause](http://docs.oasis-open.org/pkcs11/pkcs11-profiles/v2.40/os/pkcs11-profiles-v2.40-os.html#_Toc416960548) in the v2.40 specification, those functions are not within the set that all conforming implementations are required to support.

Although you can technically use this cross-sessions, [there be dragons](https://en.wikipedia.org/wiki/Here_be_dragons).

**Comment by levitte** [comment]
> ... I can't pass back the pointer to EVP_MD_CTX, but I need to copy it into the user supplied buffer. However, I don't know the size of EVP_MD_CTX because its opaque....

This is not quite true.  The user supplied buffer can simply contain the pointer to the EVP_MD_CTX or whatever, and in that case, the size of the buffer would be `sizeof(void *)`.  More cleanly, you could have the `pOperationState` point at this structure internally:

``` C
struct internal_operation_state_st {
    EVP_MD_CTX *mdctx;
};
```

and `pulOperationStateLen` would then have to be `sizeof(struct internal_operation_state_st)`.  `mdctx` would be the copy of the "original" `EVP_MD_CTX`.

Yes, of course this is a hack!

**Comment by levitte** [comment]
> Although you can technically use this cross-sessions, [there be dragons](https://en.wikipedia.org/wiki/Here_be_dragons).

It's been a while since I saw those dragons sniffing around...  I guess it was time :wink:

**Comment by simo5** [comment]
FWIW, I have knowledge of more people needing this, and it *has to be* serialized to be useful as the use case is about restoring the state in a completely new process (so _dup() is useless) within a cluster or to defer in time and let the server be re-startable.

It is perfectly fine to require same provider code version, and to set up all the other parameters explicitly, with only the Hash core internal state saved (basically whatever buffer is used to compute the next step of the hashing mechanism and ancillary variables that need to be kept around.

Allowing the same for HMACs would also be really useful.

**Comment by ifranzki** [comment]
Yes, I fully agree.

Please also note that some HSMs store the digest state in encrypted form (encrypted by the master key of the HSM) and thus the serialized state information must be allowed to have any (provider) internal format, (i.e. can be an opaque blob).

Same for HMAC states. Here the key is included, so state data is definitively sensitive and wont be exportable in clear by some HSMs.

**Comment by jimshowalter** [comment]
We have this use case, and "solved" it with bc-fips by using reflection to get at the internals of bouncy-castle, which is a horrible hack. We very much want to get off bc-fips because it's slow and always behind (it was almost three years before it supported Java 17).

Some open-source projects have a mechanism where specific bugs or features can be prioritized by paying for them. Is that true for OpenSSL?

**Comment by slontis** [comment]
@arapov FYI.

**Comment by paulidale** [comment]
In addition to the state, we'd also need to uniquely identify the provider that created the state.
Possible tractable with current infrastructure, although a provider UUID would be beneficial.

**Comment by slontis** [comment]
Some form of serialization version saved into the state may also be required.

**Comment by simo5** [comment]
> In addition to the state, we'd also need to uniquely identify the provider that created the state.
> Possible tractable with current infrastructure, although a provider UUID would be beneficial.
>

If the state is generated by the provider and is completely opaque to
the rest of OpenSSL you can leave it to the provider to detect if it
is a state it knows how to deal with.

Then it becomes similar to [de/en]coders, where you just ask in turn
the providers you know if they understand what they see.

Of course an external format that immediately identified the provider
would be a useful optimization.

**Comment by beldmit** [comment]
@vdukhovni it will require some spec (IETF-level?) but for internal purpose inside the same version we can live with smth more simple

**Comment by richsalz** [comment]
OpenSSL could use an OID of its own, and change it whenever the serialized session changes. The OID and maintenance is easy. Detecting changes is hard and requires diligence, or automation, from the developers.  Perhaps using the SONAME or its equivalent, is close enough.

**Comment by simo5** [comment]
A test can be added to CI that has a serialized state stored, if openssl fails to decode it, it is time to change the oid and regenerate the stored state

**Comment by vdukhovni** [comment]
> In addition to the state, we'd also need to uniquely identify the provider that created the state. Possible tractable with current infrastructure, although a provider UUID would be beneficial.

There's no need to serialise the provider, it is sufficient to fingerprint the *structure*, the application that resumes the hash computation may have its suite of providers. The OID identifying the data structure *may* be provider-specific, but for typical digests it should just identify the layout of the data elements, ideally allowing multiple providers to support decoding the same structure and continuing the hash computation.

**Comment by richsalz** [comment]
Agree with @vdukhovni  You absolutely DO NOT WANT to serialize the provider, that breaks interoperability.

**Comment by kgoldman** [comment]
Background on the TPM use case:

1. The hardware TPM is memory constrained, and thus has a context save/load feature to swap out resources when multiple processes share the TPM. One of those resources is hash state.

A virtual TPM wants to completely emulate the TPM, and therefore needs a serialize feature. Many vTPMs use the openssl crypto library.

2. VMs can be migrated for load balancing. To migrate its vTPM, the entire vTPM state must be serialized, including a hash operation in progress.

**Comment by pixcc** [comment]
This feature will also be useful when computing a hash over a large amount of data. Serializing the hash's intermediate state to persistent storage as a checkpoint allows the hash calculation to survive an application restarts without starting from scratch.

The feature should be compatible between different versions of OpenSSL to support seamless upgrade and rollback.

**Comment by paulidale** [comment]
How much data?  Digests are generally very fast (Gb/sec is not unheard of).

**Comment by pixcc** [comment]
> How much data? Digests are generally very fast (Gb/sec is not unheard of).

The total size of a dataset can be several petabytes. But the data is divided, so one digest per a few gigabytes. I don't think the process is bounded by the speed of hash calculation, rather you don't want to start downloading the data again - it downloads 8 megabytes at a time, but it's very parallel.

**Comment by beldmit** [comment]
From the usability perspective I don't think that hashing petabytes of data in a single piece is a good idea...

**Comment by pixcc** [comment]
> From the usability perspective I don't think that hashing petabytes of data in a single piece is a good idea...

One hash per a few gigabytes.

**Comment by t8m** [comment]
This was implemented in #28837
